### ISSUE TITLE
gh-ost hangs at `cut-over` phase at "Waiting for events up to lock"

### ISSUE THREAD

**Issue by shlomi-noach** [github/gh-ost#314]
I've had this happen twice now: running a "migrate-on-replica" migration with `postpone-cut-over-flag-file` set; migration properly waits on cut-over, but when instructing `cut-over` the following occurs:

```
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 88/100; Time: 37m0s(total), 36m37s(copy); streamer: mysql-bin.015237:917577121; State: postponing cut-over; ETA: due
2016-11-17 03:00:13 INFO executing gh-ost-on-status hook: /data/github/shell/bin/gh-ost-hooks/prod/gh-ost-on-status-hook
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 89/100; Time: 37m0s(total), 36m37s(copy); streamer: mysql-bin.015237:917607446; State: postponing cut-over; ETA: due
2016-11-17 03:00:13 INFO executing gh-ost-on-status hook: /data/github/shell/bin/gh-ost-hooks/prod/gh-ost-on-status-hook

2016-11-17 03:00:14 INFO Grabbing voluntary lock: gh-ost.1317274408.lock
2016-11-17 03:00:14 INFO Setting LOCK timeout as 6 seconds
2016-11-17 03:00:14 INFO Looking for magic cut-over table
2016-11-17 03:00:14 INFO Creating magic cut-over table `my_schema`.`_my_table_del`
2016-11-17 03:00:14 INFO Magic cut-over table created
2016-11-17 03:00:14 INFO Locking `my_schema`.`my_table`, `my_schema`.`_my_table_del`
2016-11-17 03:00:14 INFO Tables locked
2016-11-17 03:00:14 INFO Session locking original & magic tables is 1317274408
2016-11-17 03:00:14 INFO Writing changelog state: AllEventsUpToLockProcessed
2016-11-17 03:00:14 INFO Waiting for events up to lock
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 94/100; Time: 37m5s(total), 36m37s(copy); streamer: mysql-bin.015237:919839473; State: throttled, max-load Threads_running=434 >= 25; ETA: due
2016-11-17 03:00:21 ERROR critical-load met once: Threads_running=1007, >=1000. Will check again in 3000 millis
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 94/100; Time: 37m10s(total), 36m37s(copy); streamer: mysql-bin.015237:922043202; State: throttled, max-load Threads_running=784 >= 25; ETA: due
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 94/100; Time: 37m15s(total), 36m37s(copy); streamer: mysql-bin.015237:924266231; State: throttled, max-load Threads_running=773 >= 25; ETA: due
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 94/100; Time: 37m20s(total), 36m37s(copy); streamer: mysql-bin.015237:926065216; State: throttled, max-load Threads_running=670 >= 25; ETA: due
Copy: 8488615/8488615 100.0%; Applied: 10182; Backlog: 94/100; Time: 37m25s(total), 36m37s(copy); streamer: mysql-bin.015237:927794202; State: throttled, max-load Threads_running=516 >= 25; ETA: due
```

The `cut-over` begins; the `AllEventsUpToLockProcessed` gets written, but is never read back (or read but not found).
This leaves the `cut-over` phase hanging for eternity. The log above suggests an outcome of increased `threads_running` etc.

The log above also suggests the streamer is continuing to make progress, and so we have supposedly read and skipped the `AllEventsUpToLockProcessed` event.

**Comment by shlomi-noach** [comment]
Further observations:

The `AllEventsUpToLockProcessed` is known to have been written. This is evident both from the log (or the lack of error in the log) as well as actual research into the binary logs to find it there.

**Comment by shlomi-noach** [comment]
Further: in the log of another similar hang, we see:

```
Copy: 117935/117935 100.0%; Applied: 13; Backlog: 1/100; Time: 28s(total), 13s(copy); streamer: mysql-bin.014884:10486275; State: throttled, max-load Threads_running=25 >= 25; ETA: due
Copy: 117935/117935 100.0%; Applied: 13; Backlog: 1/100; Time: 29s(total), 13s(copy); streamer: mysql-bin.014884:10969584; State: throttled, max-load Threads_running=226 >= 25; ETA: due
Copy: 117935/117935 100.0%; Applied: 13; Backlog: 1/100; Time: 30s(total), 13s(copy); streamer: mysql-bin.014884:11681062; State: throttled, max-load Threads_running=328 >= 25; ETA: due
Copy: 117935/117935 100.0%; Applied: 13; Backlog: 1/100; Time: 31s(total), 13s(copy); streamer: mysql-bin.014884:12312167; State: throttled, max-load Threads_running=328 >= 25; ETA: due
```

- In both cases there is throttling involved. 

  However thrttling is disabled even before cut-over begins.

  I'll search for a race condition: throttling may have begun _just before_ cut-over begun (which is also verified beforehand, but again there may be a race condition)

- In both cases the backlog is not empty.

  Which explains why the `AllEventsUpToLockProcessed` is not processed: it's in the queue.

**Comment by shlomi-noach** [comment]
Unrelated to solving this, there should also be a timeout mechanism in [waitForEventsUpToLock](https://github.com/github/gh-ost/blob/27712325899f1c289fd9a12607636b2dcc56d551/go/logic/migrator.go#L447) that bails out for whatever reason causing a delay.

**Comment by shlomi-noach** [comment]
At this time I'm able to reliably reproduce this, and am working to fix it.

**Comment by shlomi-noach** [comment]
Found it! Indeed if throttling happens to begin just as used commands unpostpone, the cut-over blocks until throttling ends.

Am fixing.

**Comment by shlomi-noach** [comment]
Quick summary of things that have been fixed in #315:

- Previously, throttling could race up the cut-over, and would then take over and stall for as long as throttling reasons prevailed, preventing the `AllEventsUpToLockProcessed` signal from being read from the binary log, preventing `cut-over` to complete.

  This has been fixed, throttling is now ultimately non-blocking when inside `cut-over` phase.

- Furthermore, the `waitForEventsUpToLock()` now has its own timeout to bail out, so that if, for example, replication lag suddenly grows while cutting over, we break after a predetermined time and unlock the tables, further failing the `cut-over`

- `cut-over` retries was broken; it now works correctly, including postponing before `cut-over`. including throttling before cut-over, including rolling back.

**Comment by shlomi-noach** [comment]
fixed by https://github.com/github/gh-ost/pull/315
